Beta3/RC1, preprocessor information added + many fixes

juanjux opened the pull request:
NOTE: Changes will be easier to check once the beta2 PR has been merged so you can wait for that to keep your sanity when reviewing this one.

Other than fixes this version has:

- Enabled writing of all `IsWhatever` properties again, even if false so the user doesn't have to guess what properties a node can have.
- Adds preprocessor nodes. These include position information, but are not in position order in the three but hanging from the `Prop_PreprocStatements` of the root `CPPASTTranslationUnit` node. Maybe we could try to move them to their guessed position in the tree (but they could be wrong for branching statements of untaken branches since the statements will be in the AST but the contents don't, see below).
- Branching preprocessor statements (`#ifdef`, `#if`, etc) include an `IsTaken` property if that branch was choosen. Unfortunately the untaken branch nodes is not included in the AST/UAST since CDT interprets the preprocessor (but the preprocessor statements for untaken branches are present, just not the contents).
- `#include`s are tagged as semantic `Import`s and have roles added.
- Macro expansions are moved to childs of macro definitions (they're separate lists from CDT).
- All nodes that are expanded from a macro have a `ExpandedFromMacro` property. The value of this property is the string `{macroName}_{macroDefinitionStartOffset}_{macroDefinitionEndOffset}`. Alternatively we could also tag each macro definition with some hash and use that hash in the expanded nodes, but I think this is clearer.

TO-DO:

- [X] `FunctionGroup` varargs fix.
- [X] Add roles to other preprocessor statements.
- [X] Fix an issue that arises with the latest Anna's test.
- [X] `QualifiedIdentifier` semantic objects [blabla N/A].
- [x] On semantic functions with a `QualifiedIdentifer` name, join the parts to produce the expected single `Identifier`.
- [X] Rebase.

Could be RC1 after these fixes.

To think about: 

Because of the preprocessor branches, this driver could work a lot better for the users if they could pass parameters (like the compile `-Dwhatever` option) so they can get the expected AST for their parameters (like `-D_WIN32`, `-Dunix`, etc). So we probably should evaluate how to do it. One option would be to add  `SetParameter(key, val)` or `SetParameters(dict)` calls to the protocol. Or `ParseWithParameters(...)`. Or forwarding some environment variables (from the client to bblfshd to the drivers...).

dennwc commented:
Needs a rebase

juanjux commented:
@dennwc you're too fast I was in the middle of it :D

dennwc commented on .idea/cpp-driver.iml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file should not be here

juanjux replied:
Done and added to .gitignore.

dennwc commented on driver/normalizer/normalizer.go:
```diff
@@ -90,26 +103,34 @@ var Normalizers = []Mapping{
-		},
-	)),
+	// Disabled: parts can be anything, not only identifiers (like function calls or object instantiations)
+	// and the SDK always expects identifiers (but should except Any[] for this object)
```
As discussed in Slack, `QualifiedIdentifier` in SDK works as intended. The one in C++ needs to be split into `QualifiedIdentifier` and "selector expressions" similar to [this](https://github.com/bblfsh/go-driver/blob/master/fixtures/calls.go.native#L904).

juanjux replied:
I've noticed that all the parts of a `QualifiedName` in the native AST implement the interface `IASTName` so with a somewhat huge `Case` it should be possible to get all the name parts (working on it right now).

juanjux replied:
Fixed in next commit, but this creates the new problem that `FunctionGroup.Name` expect an Identifier and sometimes the value is a `QualifiedIdentifier` so I'll need to join the `QualifiedIdentifier.Names` into a single string to pass as a single `Identifier` (if there is a better solution please tell me).

dennwc replied:
Hmm, you are right, I haven't expected an `Alias` to be a `QualifiedIdentifier`. I'll check it.

dennwc commented on driver/normalizer/normalizer.go:
```diff
@@ -90,26 +103,34 @@ var Normalizers = []Mapping{
 		Fields{
-			//{Name: "IsDefaulted", Optional: "optDefaulted", Op: Var("ignDefaulted")},
-			//{Name: "IsDeleted", Optional: "optDeleted", Op: Var("ignDeleted")},
+			{Name: "IsDefaulted", Optional: "optDefaulted", Op: Var("ignDefaulted")},
```
As you know, I'm not a fan of dropping fields :)

There are only three valid options here:
1) You expect the field to always be `nil`/`false`, so you can use `Is(nil)` or `Bool(false)`.
2) You know that the field value varies, but don't know what to do with it. Unfortunately we don't have every feature defined in Semantic UAST, so for now I think it's okay to make a fake node in `FunctionGroup` and dump those values to it. But please **don't drop fields** like this.
3) The node is a complete garbage and just duplicates the info in the main node (happens with JS and Python). Then it's the only one case when you can drop it with `Any(nil)` operation.

juanjux replied:
I personally don't like any of these options:

1. They aren't always nil/false, they can be true or false.
2. This is the case with these fields. I don't think dumping them into a fake node in the `FunctionGroup` will be of much help (or pretty). Also, in this AST they're not of a single node, but of almost all nodes that form the `FunctionGroup`: each argument, return values, etc. In this case, we would need many fake nodes, which is even uglier.
3. Not the case.

In my opinion, dropping fields that doesn't have a representation in the defined schema of the semantic object is not bad. The user always can go back to the non-semantic UAST or even the native one to get those. We knew that making a common denominator in semantic objects was going to lose information on some languages.

But I propose that every node in the semantic UAST could have a child node of type `Extra` (which would be in the schema of all semantic nodes) or something like that storing all the properties that doesn't conform to the parent node schema. This way we don't lost anything but we don't pollute the main group with fake nodes at unrelated locations.

PS: let's continue the discussion here: https://github.com/bblfsh/sdk/pull/334#pullrequestreview-180196127

juanjux replied:
As discussed on the other PR, I changed those fields to use `AnyNode(nil)` to drop them until we decide on a better solution in the SDK for these unmatched properties.

dennwc commented on driver/normalizer/normalizer.go:
```diff
@@ -119,63 +140,78 @@ var Normalizers = []Mapping{
-					//{Name: "IsTypeName", Op: Var("ignIsTypeName")},
-					{Name: "StorageClass", Op: Var("StorageClass")},
-					{Name: "IsVirtual", Optional: "optIsVirtual", Op: Var("ignIsVirtual")},
+					{Name: "IsComplex", Optional: "optIsComplex", Op: Var("ignIsComplex")},
```
If I understood correctly, those fields will always appear now, right? If so, please remove `Optional` everywhere.

juanjux replied:
I'm on it, but some are still really `Optional` because they can appear or not (mostly properties contained children aka the `Prop_*`s).

For the others, I'll use a `Case`. It's more cumbersome than one or two `Optional`s on simple cases but more correct I guess.

dennwc commented on fixtures/_integration.cpp.sem.uast:
```diff
@@ -208,6 +225,27 @@
    ],
-   'Prop_IncludeDirectives': [],
+   'Prop_PreprocStatements': [
+      { '@type': "uast:Import",
```
Specifically `uast:InlineImport`... That is commented out in SDK :D Will uncomment it back.

juanjux replied:
Ping me here when its done and I'll change it.

juanjux replied:
Noticed the PR :) merged, tagged, dependency upgraded and changed.

creachadair commented on driver/fixtures/fixtures_test.go:
```diff
@@ -28,7 +28,7 @@ var Suite = &fixtures.Suite{
 			"CPPASTName",
 			"CPPASTCompoundStatement",
-			"CPPASTQualifiedNames",
+			//"CPPASTQualifiedNames",
```
Is this meant to be deleted?

juanjux replied:
Is currently disabled, see discussion above with Dennys above what to do with these (the schema expects a list of `Identifiers` but in this driver, and most really, the parts can be other nodes like function calls).

juanjux replied:
Fixed!

creachadair commented on driver/normalizer/annotation.go:
```diff
@@ -136,15 +136,22 @@ var Annotations = []Mapping{
+	AnnotateType("CPPASTImplicitName", FieldRoles{
+		"Name": {Rename: uast.KeyToken},
+	}, role.Identifier),
+	AnnotateType("ASTInclusionStatement", FieldRoles{"Name": {Rename: uast.KeyToken}}, role.Import),
```
```suggestion

	AnnotateType("ASTInclusionStatement", FieldRoles{
		"Name": {Rename: uast.KeyToken},
	}, role.Import),
```

for consistency with the others.

juanjux commented:
Fixed the `QualifiedIdentifier` annotation and missing unannotated preprocessor roles plus all the reported problems in the review (thanks, people :heart: ).

There is still a pending problem because native `Function` names can be `Identifiers` or `QualifiedIdentifiers` but only the former are expected in the semantic schema so I need to join the parts of the qualified ones in a single `Identifier`. This will make the build fail and some fixtures to be missing until fixed.

This also needs a rebase before merging to un-pollute the history without the gazillion commits of this PR.

juanjux commented:
Fixed the last problem, the case when the value of a `FunctionGroup.Alias.Name` was a `QualifiedIdentifier` instead of an `Identifier` by adding an `Op` that will join the elements. Not reversible, trough, I hope to fix that with some help from @dennwc.

juanjux commented:
Rebased to a smaller and more differentiated collection of commits, removed old cruft and moved one inner class to their own file so the visitor class is slightly shorter now.

Ready for another review round @creachadair , @dennwc , @bzz .

dennwc commented on driver/normalizer/annotation.go:
```diff
@@ -235,12 +259,12 @@ var Annotations = []Mapping{
 		uast.KeyToken: Var("operator"),
 		"Prop_Operand1": ObjectRoles("operand1", role.Binary, role.Expression, role.Left),
 		"Prop_Operand2": ObjectRoles("operand2", role.Binary, role.Expression, role.Right),
+		"Prop_InitOperand2": ObjectRoles("init_operand2", role.Binary, role.Expression, role.Right),
```
It's also a `Binary,Expression,Right`? It should probably be either `Incomplete` or `Initialization` as well.

juanjux replied:
Yeah. It's that node that was always the same as Prop_Operand2 except in one of the bench_ tests where is slightly different (remember that I propopsed to remove it in the previous PR). Adding those roles in the next commit.

dennwc commented on driver/normalizer/normalizer.go:
```diff
@@ -27,7 +28,78 @@ var Preprocessors = []Mapping{
+}
+
+func (op opJoinNamesArray) Kinds() nodes.Kind {
+	return nodes.KindString
```
It should probably return the same value as `op.qualified` does.

juanjux replied:
What do you mean? The op return a string (the joined identifiers).

dennwc replied:
Ah, yeah, confused with the `Check` of the source.

dennwc commented on driver/normalizer/normalizer.go:
```diff
@@ -27,7 +28,78 @@ var Preprocessors = []Mapping{
+}
+
+func (op opJoinNamesArray) Check(st *State, n nodes.Node) (bool, error) {
+	res, err := op.qualified.Check(st, n)
```
Just `return op.qualified.Check`.

Also, this seems to be an invalid implementation. It should do a reverse of `Construct`: expect a string, split it, generate nodes and pass to `op.qualified.Check`.

juanjux replied:
It wasn't clear to me where to put the reverse transformation but until confirming with you I suspected it was there :) I'll add it.

juanjux replied:
I just commited this. Please check because I'm not sure I did everything as it should and I also don't know how to check the result of the reverse operation.

dennwc replied:
Looks correct, thanks! The `uast.ToNode` is slow, in general, but since we won't use reverse that often, I think it's okay.

The best way is to write a small test for it. Should be something like:
```golang
op := opJoinNamesArray{Var("x")}

st := NewState()
n2, err := op.Check(st, n1)
// test that n2 should is a specific string

// reverse
n3, err := op.Construct(st, n2)
// test that it returned an array of objects with names
```

juanjux replied:
Trying to get a test to work but getting very odd results (even in the case where it clearly works on the fixtures) :/

dennwc commented on driver/normalizer/normalizer.go:
```diff
@@ -27,7 +28,78 @@ var Preprocessors = []Mapping{
+		Obj{
+			"Path":  UASTType(uast.Identifier{}, Obj{"Name": Var("path")}),
+			"All":   Bool(true),
+			"Names": Arr(),
```
`Is(nil)` should be ok.

dennwc commented on driver/normalizer/normalizer.go:
```diff
@@ -73,118 +145,183 @@ var Normalizers = []Mapping{
 	// Args in C can have type but be empty (typically in headers, but also in implementations): int main(int, char**)
```
Those nodes should be removed completely. The corresponding `Argument` should have `Name: nil`.

juanjux replied:
It's not only for functions. If I removed this then some `CPPASTName`s wouldn't be promoted to `uast:Identifier` (and cause the test to fail) because there are other empty instances of `CPPASTName`s in non semantic objects like enums and other cases. I don't think an empty identifier is against the specs. Or is there any way to promote all empty `CPPASTName`'s to a nil in a way that the `nil` var will be used everywhere the previous value was expected?

dennwc replied:
I guess mapping from a node to `Is(nil)` should work well. But it will leave nil nodes in arrays (if Identifiers ever happen to be in arrays).

juanjux replied:
You mean like this?

```go
	Map(Obj{
		"IASTClass": String("CPPASTName"),
		"Name":      String(""),
	}, Obj{
		uast.KeyType: String("uast:Identifier"),
		"Name":       Is(nil),
	}),
```

Or this?

```go
	Map(Obj{
		"IASTClass": String("CPPASTName"),
		"Name":      String(""),
	},
		Is(nil),
	),
```

juanjux replied:
Second option seems to work well and doesn't break anything _except_ one seemingly unrelated annotation in one fixture removing the roles :-? Investigating it.

juanjux replied:
Fixed :)

dennwc commented on driver/normalizer/normalizer.go:
```diff
@@ -73,118 +145,183 @@ var Normalizers = []Mapping{
+	MapSemantic("CPPASTQualifiedName", uast.QualifiedIdentifier{}, MapObj(
 		Obj{
-			"Prop_AllSegments": Var("names"),
+			"Prop_AllSegments": Each("qualParts", Cases("caseQualParts",
```
This defines multiple cases, but they are not used below.
I guess I'll need to implement "unused variable" check for state variables as well :stuck_out_tongue:

juanjux replied:
Mmmm, outside of the `qualPos` that I forgot to remove (that certainly should have been detected, will be removed in the next commit) I think having several cases on one side extracting a single `Var` (`name` in this case) and not requiring several cases on the other side when a single case just use that var looks more like a feature than a bug! (otherwise we would have 3 exactly equal cases on the right/low side).

dennwc replied:
I agree that it sounds like a good feature. Although reverse transform will be problematic in this case. And having 3 copies in the destination transform won't help.

So yeah, let's leave it as-is, but please add a comment to the destination transform that the case variable is intentionally omitted.

juanjux replied:
Done in the next commit.

dennwc commented on fixtures/_integration.cpp.sem.uast:
```diff
@@ -208,6 +176,27 @@
+         All: true,
+         Names: [],
+         Path: { '@type': "uast:Identifier",
+            Name: "iostream",
```
Will it distinguish `<iostream>` and `"iostream"`?

juanjux replied:
Not in the semantic UAST, in the normal UAST there is a `IsSystem` boolean that discriminates between both.

dennwc replied:
I would suggest prepending `./` if `IsSystem == false`. This will mirror the way how other drivers do it.

juanjux replied:
Another option would be including the quotes or the brackets in the name, but that could give parsing headaches for the user so I'll go with your suggestion.

juanjux replied:
Done!

dennwc commented on fixtures/_integration.cpp.sem.uast:
```diff
@@ -208,6 +176,27 @@
+         },
+         All: true,
+         Names: [],
+         Path: { '@type': "uast:Identifier",
```
It's probably `uast:String`, not `uast:Identifier` (since it's a path).

juanjux replied:
Would be a little odd to tag it as a string when sometimes there aren't quotes (`<iostream>`).

dennwc replied:
By the definition, Identifiers are names of objects. `iostream` is more a filename/path and other drivers express it as a `String` instead of `Identifier`. As mentioned above, there is a different way to represent `<` (`IsSystem`).

juanjux replied:
Done in next commit.

dennwc commented on fixtures/bench_accumulator_factory.cpp.sem.uast:
```diff
@@ -0,0 +1,3050 @@
+                                    Name: ~,
+                                    Receiver: false,
+                                    Type: { '@type': "uast:Identifier",
+                                       Name: "unespecified",
```
Looks like a void return. In this case, `Returns` should be `nil`.

juanjux replied:
It's a constructor so it doesn't have a explicit `void` return (but implicitly it does, of course), but yes, `Returns` should be `nil` in this case, will change it.

juanjux replied:
Done

dennwc commented on fixtures/bench_binary_search.cpp.sem.uast:
```diff
@@ -0,0 +1,2252 @@
+                                                                  '@pos': { '@type': "uast:Positions",
+                                                                     start: { '@type': "uast:Position",
+                                                                        offset: 143,
+          
```
Wow. I wonder what the hell is this :grin: Maybe calling string conversion method on some complex structure?

juanjux replied:
It's really a not so complex function body, but this driver is very verbose (even after several liposuctions in our side).

dennwc replied:
Damn, Github is doing something weird with comments. It should be in a totally different place:
https://github.com/bblfsh/cpp-driver/pull/16/files/3f6092d41862e714e25c5ab2466b244ea8a00ed1#diff-c79cc6306ed8f76da214a1657ea888efR1076

juanjux replied:
Ahhh, the funny catch-all `ExpressionType` field. It's some kind of signature the native AST generates for every expression. For simple ones it's very nice since it gives the inferred type (like `const char[14]`), for others is some beauty like this. When there is a parse error it'll also add some string indicating the error in that field.

dennwc commented:
Latest changes look good! Only the relative include and the reverse transform is left.

juanjux commented:
@dennwc only the reverse transformation now, working on it.

![](https://www.catholicgentleman.net/wp-content/uploads/2013/10/but-i-can-carry-you-lord-of-the-rings-2148649-600-833.jpg)

juanjux commented:
@dennwc fixed the reverse transformation (with your help) and added a test, PTAL.

dennwc commented on driver/normalizer/normalizer.go:
```diff
@@ -64,6 +64,25 @@ func (op opJoinNamesArray) Kinds() nodes.Kind {
+		qual.Names = append(qual.Names, id)
+	}
+
+	n, err := uast.ToNode(qual.Names)
```
Hmm, interesting. You don't need `qual` here, actually. You can declare `var names nodes.Array` and `ToNode(id)` for each element, or maybe `[]Identifier` or `[]Any` will work as well.

juanjux replied:
Changed to `[]Identifier`.

dennwc commented on driver/normalizer/normalizer_test.go:
```diff
@@ -1,23 +1,48 @@
+	n4, ok := n3.(nodes.String)
+	require.True(t, ok)
+	require.Equal(t, n1, n4)
+	require.Equal(t, string(n4), "a::b::c")
```
Nit: "expected" value goes first

juanjux replied:
Changed.

juanjux commented:
@dennwc nitpickings fixed

creachadair commented:
Woot!

Thanks @juanjux for sticking with this tricky change, and @dennwc for the close and careful review!
